### PR TITLE
refactor: create branch cache before processing branch

### PR DESCRIPTION
**Pull request by RahulGautamSingh**
## Changes
- branchCache renamed to branchState in `process/write`
- we now create a branchCache even when repositoryCache is disabled OR branchExisted is false

## Context
- Part of #17538
- BranchCache term can get confusing when we use it while the repositoryCache is disabled. So we now use branchState instead, which is almost same logic wise but is easier to understand. Genreal idea is that we have a branchState which we derive initially from the cache and then use it during the execution phase. Finally when setting the cache we use the update values from branchState.

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/util/git/modified-cache.ts [hunk 1]
```diff
@@ -24,12 +24,12 @@ export function setCachedModifiedResult(
...
   const { branches } = cache;
-  const branch =
-    branches?.find((branch) => branch.branchName === branchName) ??
-    ({ branchName: branchName } as BranchCache);
+  let branch = branches?.find((branch) => branch.branchName === branchName);
 
   // if branch not present add it to cache
-  if (is.undefined(branch.sha)) {
+  if (!branch) {
+    logger.warn(`Branch cache not present for ${branchName}`);
```
Comment (rarkins) [inline lib/util/git/modified-cache.ts:31]: This should never happen?

Comment (RahulGautamSingh) [reply lib/util/git/modified-cache.ts:31]: Yes now it won't wasn't sure whether to warn or use thee old code, need suggestion

Comment (RahulGautamSingh) [reply lib/util/git/modified-cache.ts:31]: In fact, I was able to locate an instance of this. It occurs while eliminating stale branches, thus it's a good idea not to create new branch caches here as doing so will cause us to store an extra branch during the cleanup cycle. 

How to reproduce:
1. Create a standard repository with dependencies set between 0.0.0 the most recent version.
2. Run renovate -> update PR created
3. Adjust the version number in the base branch to greater most recent version. (if most recent is 1.2.0 update it to 1.2.3)
4. Run renovate once more 

Result:
image

File: lib/util/cache/repository/types.ts [hunk 2]
```diff
@@ -28,13 +28,16 @@ export interface BranchUpgradeCache {
...
   automerge: boolean;
+  branchFingerprint?: string;
   branchName: string;
   isModified: boolean;
+  /**
+   * Parent commit of branch sha (latest branch commit)
+   */
+  parentSha: string | null;
   prNo: number | null;
   sha: string | null;
```
Comment (viceice) [inline lib/util/cache/repository/types.ts:39]: can you add a tsdoc comment to the other fields too, so it's clear what they are for?

File: lib/util/git/modified-cache.spec.ts [hunk 3]
```diff
@@ -7,6 +8,8 @@ import {
 } from './modified-cache';
 
 jest.mock('../cache/repository');
+jest.mock('../../logger');
+const logger = mocked(_logger);
```
Comment (viceice) [inline lib/util/git/modified-cache.spec.ts:12]: logger is globally mocked, so no need to do it again .

mocked logger can be imported from test/util

File: lib/util/cache/repository/types.ts [hunk 4]
```diff
@@ -27,14 +27,38 @@ export interface BranchUpgradeCache {
...
+  /**
+   * Branch name
+   */
   branchName: string;
+  /**
+   * Whether a person not listed in gitIgnoredAuthors updated the branch.
+   */
   isModified: boolean;
+  /**
+   * Parent commit of branch sha
```
Comment (viceice) [inline lib/util/cache/repository/types.ts:47]: is this the commit Sha when the branch was created from base branch?

File: lib/util/git/modified-cache.ts (hunk 1 above)
Comment (rarkins) [reply lib/util/git/modified-cache.ts:29]: I don't think we should warn then, if it's possible normal behavior

File: lib/util/cache/repository/types.ts (hunk 4 above)
Comment (RahulGautamSingh) [reply lib/util/cache/repository/types.ts:47]: Yes and no.
When the branch was first created "yes". But if the branch was modified after that then it is the 2nd last commit.
Ie. 
Branch Created and has sha as S1, parentSha as P1
Now user or bot makes another commit, sha becomes S2 and parentSha becomes S1.

Comment (rarkins) [reply lib/util/cache/repository/types.ts:47]: What's the reason for *updating* this over time?

File: lib/util/git/modified-cache.ts [hunk 5]
```diff
@@ -9,11 +8,11 @@ export function getCachedModifiedResult(
   const { branches } = getCache();
   const branch = branches?.find((branch) => branch.branchName === branchName);
 
-  if (branch?.sha !== branchSha) {
-    return null;
+  if (branch?.sha && branch.sha === branchSha) {
+    return branch.isModified;
```
Comment (rarkins) [inline lib/util/git/modified-cache.ts:12]: Should we check if branch.isModified is boolean?

File: lib/util/git/modified-cache.ts [hunk 6]
```diff
@@ -22,18 +21,16 @@ export function setCachedModifiedResult(
...
-    ({ branchName: branchName } as BranchCache);
+  const branch = cache.branches?.find(
+    (branch) => branch.branchName === branchName
+  );
 
-  // if branch not present add it to cache
-  if (is.undefined(branch.sha)) {
-    branches.push(branch);
+  if (!branch) {
+    logger.debug(`Branch cache not present for ${branchName}`);
```
Comment (rarkins) [inline lib/util/git/modified-cache.ts:29]: Will this log every time when repoCache is not enabled?

File: lib/workers/repository/process/write.ts [hunk 7]
```diff
@@ -59,22 +60,30 @@ export async function writeUpdates(
...
-    if (branchExisted && config.repositoryCache === 'enabled') {
-      branchCache =
-        cachedBranches?.find((br) => br.branchName === branchName) ??
-        ({} as BranchCache);
+    let branchState =
+      cachedBranches.find((br) => br.branchName === branchName) ??
+      ({} as BranchCache);
 
-      if (Object.keys(branchCache).length === 0) {
+    if (Object.keys(branchState).length === 0) {
```
Comment (rarkins) [inline lib/workers/repository/process/write.ts:74]: This logic seems unnecessarily complicated. First you initialize it as `{}` and then immediately after you check whether it's empty. Easier to just not initialize it and check if the value is falsey.

**Review by rarkins — CHANGES_REQUESTED** [review]

File: lib/util/git/modified-cache.ts (hunk 6 above)
Comment (RahulGautamSingh) [reply lib/util/git/modified-cache.ts:29]: No. A minimal branchState is created always which will prevent this.

File: lib/workers/repository/process/write.ts (hunk 7 above)
Comment (RahulGautamSingh) [reply lib/workers/repository/process/write.ts:74]: I had updated it but forgot to push that code I think, my bad😅

File: lib/util/cache/repository/types.ts (hunk 4 above)
Comment (RahulGautamSingh) [reply lib/util/cache/repository/types.ts:47]: Not sure what you mean to ask.

Comment (rarkins) [reply lib/util/cache/repository/types.ts:47]: If there's a new commit, what's the value/reason for updating `parentSha` to value `S1`? Do we use that value anywhere?

Comment (RahulGautamSingh) [reply lib/util/cache/repository/types.ts:47]: You metioned that you use it in the app-backend to compute mergeConfidence

Comment (RahulGautamSingh) [reply lib/util/cache/repository/types.ts:47]: Also it gets calculated likt this each run:
```
 parentSha = await git.revparse([`${branchSha!}^`]);
```
So if we want to use cached value on re-runs we need to update its value in cache. Else while setting cache we will end up cloning again.

File: lib/workers/repository/process/write.ts [hunk 8]
```diff
@@ -59,22 +60,28 @@ export async function writeUpdates(
 
   for (const branch of branches) {
     const { baseBranch, branchName } = branch;
+    const branchSha = getBranchCommit(branchName)!;
```
Comment (viceice) [inline lib/workers/repository/process/write.ts:63]: Does this cause a clone?

File: lib/util/git/modified-cache.ts [hunk 9]
```diff
@@ -9,11 +8,15 @@ export function getCachedModifiedResult(
   const { branches } = getCache();
   const branch = branches?.find((branch) => branch.branchName === branchName);
 
-  if (branch?.sha !== branchSha) {
-    return null;
+  if (
+    branch?.sha &&
+    branch.sha === branchSha &&
```
Comment (viceice) [inline lib/util/git/modified-cache.ts:13]:
```suggestion
    branch?.sha === branchSha &&
```

isn't this the same? so if `sha` is missing, `isModified` is missing too? 🤔

File: lib/util/cache/repository/types.ts (hunk 4 above)
Comment (viceice) [reply lib/util/cache/repository/types.ts:47]: ok, so this is wrong in all cases when there are commits ontop of the renovate branch?
I think we need to fix this too. it should probably always point to the renovate commit (eg first branch commit)

File: lib/workers/repository/process/write.ts (hunk 8 above)
Comment (RahulGautamSingh) [reply lib/workers/repository/process/write.ts:63]: No

File: lib/util/cache/repository/types.ts (hunk 4 above)
Comment (RahulGautamSingh) [reply lib/util/cache/repository/types.ts:47]: Not sure. @rarkins thoughts??

Comment (viceice) [reply lib/util/cache/repository/types.ts:47]: @RahulGautamSingh please create an issue to track this, so we can discuss it separately

**Review by viceice — APPROVED** [review]

**Review by rarkins — APPROVED** [review]

**Pull request merged**